src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -0,0 +1,427 @@
+    RetryLeasingWorkerFromNode(actor, node);
+  }
+}
+
+void GcsActorScheduler::RetryLeasingWorkerFromNode(
+    std::shared_ptr<GcsActor> actor, std::shared_ptr<rpc::GcsNodeInfo> node) {
+  auto node_id = ClientID::FromBinary(node->node_id());
+  execute_after(io_context_,
+                [this, node_id, node, actor] {
+                  auto iter = node_to_actors_when_leasing_.find(node_id);
+                  if (iter != node_to_actors_when_leasing_.end()) {
+                    // If the node is still available, the actor must be still in the
+                    // leasing map as it is erased from leasing map only when
+                    // `CancelOnNode` or the `RequestWorkerLeaseReply` is received from
+                    // the node, so try leasing again.
+                    RAY_CHECK(iter->second.count(actor->GetActorID()) != 0);
+                    RAY_LOG(INFO) << "Retry leasing worker from " << node_id
+                                  << " for actor " << actor->GetActorID();
+                    LeaseWorkerFromNode(actor, node);
+                  }
+                },
+                RayConfig::instance().gcs_lease_worker_retry_interval_ms());
+}
+
+void GcsActorScheduler::HandleWorkerLeasedReply(
+    std::shared_ptr<GcsActor> actor, const ray::rpc::RequestWorkerLeaseReply &reply) {
+  const auto &retry_at_raylet_address = reply.retry_at_raylet_address();
+  const auto &worker_address = reply.worker_address();
+  if (worker_address.raylet_id().empty()) {
+    // The worker did not succeed in the lease, but the specified node returned a new
+    // node, and then try again on the new node.
+    RAY_CHECK(!retry_at_raylet_address.raylet_id().empty());
+    actor->UpdateAddress(retry_at_raylet_address);
+    auto actor_table_data =
+        std::make_shared<rpc::ActorTableData>(actor->GetActorTableData());
+    // The backend storage is reliable in the future, so the status must be ok.
+    RAY_CHECK_OK(actor_info_accessor_.AsyncUpdate(actor->GetActorID(), actor_table_data,
+                                                  [this, actor](Status status) {
+                                                    RAY_CHECK_OK(status);
+                                                    Schedule(actor);
+                                                  }));
+  } else {
+    // The worker is leased successfully from the specified node.
+    std::vector<rpc::ResourceMapEntry> resources;
+    for (auto &resource : reply.resource_mapping()) {
+      resources.emplace_back(resource);
+    }
+    auto leased_worker = std::make_shared<GcsLeasedWorker>(
+        worker_address, std::move(resources), actor->GetActorID());
+    auto node_id = leased_worker->GetNodeID();
+    RAY_CHECK(node_to_workers_when_creating_[node_id]
+                  .emplace(leased_worker->GetWorkerID(), leased_worker)
+                  .second);
+    actor->UpdateAddress(leased_worker->GetAddress());
+    CreateActorOnWorker(actor, leased_worker);
```

Comment:
Should this flush the worker address (still with state PENDING) to the actor table before `CreateActorOnWorker`?

Reply:
> Should this flush the worker address (still with state PENDING) to the actor table before `CreateActorOnWorker`?

There is no need to flush here, because if the GCS Server crash and restarts, it will definitely lease a worker from the previously recorded node which return the previously leased worker if it already leased before.

Reply:
Hmm I'm actually not sure if the node manager guarantees that it will return the previously leased worker. The current implementation looks like it would just drop the second lease request because it already has a "task" with the same ID queued. Can you either do the additional flush or add a TODO to handle this case at the node manager?

Reply:
> Hmm I'm actually not sure if the node manager guarantees that it will return the previously leased worker. The current implementation looks like it would just drop the second lease request because it already has a "task" with the same ID queued. Can you either do the additional flush or add a TODO that we should handle this case once we try and support persistent GCS?

In fact, the logic on `NodeManager` and `CoreWorker` that you asked me to delete in the last two days is not only the logic handling the recovery of GCS Server, but also the logic handling the RetryLeasing and RetryCreating message from GCS Server. The deleted logic on `NodeManager` is used to record which worker has been leased by which actor. The logic of `CoreWorker` deletion is to record whether the worker has received the create actor request. If the two logics are still  in `NodeManager` and `CoreWorker`, there is no need to flush here. 🙂

Reply:
Ah I see, thanks for the explanation. Well, anyway, I think it's best we leave this to a future PR since this one shouldn't deal with persistence yet. I lean more towards doing the Flush here instead of trying to reason about duplicate lease requests at the `NodeManager` level (if we do that, we'll have to somehow check that it is indeed because of this scenario and not because of some other bug), but we can discuss once this PR is merged.